### PR TITLE
Logging feature

### PR DESCRIPTION
**Pull request by jychoi-hpc**
Add a python logging feature (`https://docs.python.org/3/library/logging.html`) to print messages to the screen as well as to a file (`run.log`).

Three functions are added:
* `logging_init`: setup logging
* `info`: print message prefixed with "[rank:INFO]"
* `log`: print message prefixed with "[rank:DEBUG]"

### CODE REVIEW

File: hydragnn/utils/print_utils.py [hunk 1]
```diff
@@ -63,3 +67,42 @@ def iterate_tqdm(iterator, verbosity_level):
...
+    handlers.append(logging.FileHandler(fname))
+    logging.basicConfig(level=logging.DEBUG, format=fmt, handlers=handlers)
+
+
+"""
+uage: info (str1, str2, ...). Use just like print. "[rank:INFO]" will be prefixed.
+"""
+
+
+def info(*args, logtype="info", sep=" "):
```
Comment (streeve) [inline hydragnn/utils/print_utils.py:94]: Not sure I follow the difference between info and log

**Review by streeve — COMMENTED** [review]
Looks very nice! Looks like real errors in the CI though

File: hydragnn/utils/print_utils.py (hunk 1 above)
Comment (jychoi-hpc) [reply hydragnn/utils/print_utils.py:95]: The difference is logging level. `info()` will tag outputs with "INFO" and log will tag outputs with "DEBUG". Later we can make an option to suppress "DEBUG" messages for cleaner output.

Comment (streeve) [reply hydragnn/utils/print_utils.py:95]: How about `log()` "LOG" and `debug()` "DEBUG". Unless what you're suggesting is standard

File: hydragnn/run_training.py [hunk 2]
```diff
@@ -60,6 +60,9 @@ def _(config: dict):
 
     world_size, world_rank = setup_ddp()
 
+    prefix = get_log_name_config(config)
+    logging_init(prefix, world_rank)
```
Comment (streeve) [inline hydragnn/run_training.py:64]: `setup_log` to match `setup_ddp` a little more?

Comment (jychoi-hpc) [reply hydragnn/run_training.py]: Yes it makes sense. I have changed.

File: hydragnn/utils/print_utils.py (hunk 1 above)
Comment (jychoi-hpc) [reply hydragnn/utils/print_utils.py]: I have changed too. `into()` to "LOG()", `debug()` to "DEBUG()"

**Comment by streeve** [comment]
I read more about the `logging` package and understand a little better. Now I see why you named them `info` and `debug`, but I'm hoping we can integrate this with `print_distributed` a little more. I would like to able to print debug, general info, and warnings (maybe errors?) with a given verbosity throughout the code - do you see a clean way to do so? If not, maybe we just use the info level wrapped inside as you have now

**Comment by streeve** [comment]
That sounds great to me

File: hydragnn/utils/print_utils.py [hunk 3]
```diff
@@ -63,3 +67,39 @@ def iterate_tqdm(iterator, verbosity_level):
         return tqdm(iterator)
     else:
         return iterator
+
+
+"""
+Setup logging to print messages for both screen and file.
```
Comment (streeve) [inline hydragnn/utils/print_utils.py:73]: I think if you move the docstrings in the functions `black` will stop forcing these extra blank lines

File: hydragnn/utils/print_utils.py (hunk 1 above)
Comment (streeve) [reply hydragnn/utils/print_utils.py]: Sorry for the confusion I caused. Like you say below, functions for `debug()`, `info()`, and `warning()` work for me

**Comment by jychoi-hpc** [comment]
I found adding the concept of log level (debug, info, warning) needs more discussion among us (needs some changes in many places). Instead, I figured out a clean way without breaking the current code.

The logging is same as before. Now, we can print both screen and save to `run.log` at the same time.

**Review by streeve — APPROVED** [review]
I like this a lot - it does exactly what we need. I tested locally and added to the examples as well

**Comment by streeve** [comment]
I okay to squash merge this once someone else approves

**Review by allaffa — APPROVED** [review]

**Comment by streeve** [comment]
Ready to merge after fixing issues with python 3.7

**Comment by jychoi-hpc** [comment]
Made modification to work with python 3.7 too. I tested this working with python 3..7, 3.8, and 3.9.

**Review by pzhanggit — APPROVED** [review]

**Pull request merged**